rfcs/007 - Engine tweaks.md

```diff
@@ -0,0 +1,113 @@
+
+#### ECS integration tests
+
+The problem in ecs. We were calling from the `TaskDefinition`'s `cloud.update` function the `db.create` function to force the creation of a new task definition. Again, the filter in the `apply` loop is not seeing this change because it was not done by the user and it was done after the commit started.
+
+#### Temporal solution
+
+Both test suites are passing now because we were able to add the right logic in each mapper to handle this, but somehow it's breaking module encapsulation which we want to keep at a minimum or inexistent.
+
+### High-level steps of the current `commit` algorithm
+
+1. Get relevant audit logs
+2. Get relevant modules based on step 1
+3. If any change:
+    1. Execute partial `apply` (passing only the relevant changes and modules)
+    2. If error on partial `apply`, execute full `apply`
+4. Execute full `sync`
+5. If full `apply` or `sync` error, execute `revert`
+6. Return the `sync` response
+
+### High-level steps of the current `apply` algorithm
+
+Outer loop:
+1. Execute `db.read`s
+2. Recreate changes after the commit started
+
+    Inner loop:
+    1. Execute `cloud.read`s
+    2. If necessary, recreate relevant changes
+    3. Execute `diff`ing logic
+    4. If relevant changes, filter them
+    5. If no relevant changes, exclude changes done after `commit` started
+    6. Mappers execution
+
+### High-level steps of the current `sync` algorithm
+
+Outer loop:
+1. Execute `cloud.read`s
+
+    Inner loop:
+    1. Execute `db.read`s
+    2. Recreate changes after the commit started
+    3. Execute `diff`ing logic
+    5. Exclude changes done after `commit` started
+    6. Mappers execution
+
+## Proposal
+
+TBD
+
+## Alternatives Considered
+
+It's tricky to find a common solution for both cases found. For the ECS case might be simpler, we need to find a way to also see the changes done by the engine during the `commit`. In the EC2 case, we need either to ignore the relevant changes and `apply` for all modules or find a way to "mark" entities that might be affected by another entity change and somehow include them in the filter.
+
+### Executing "full" `apply`
+This is one possible solution that could solve both issues. Would be similar to the current algorithm but we do not get "relevant changes". We always `apply` for all modules and only exclude changes done **by the user** after `commit`.
+
+#### Pros 
+- Solves the 2 issues
+- Easier to implement?
+
+#### Cons
+- We lose the "faster" `apply` since we need to iterate through all modules installed and not only the ones with changes.
```

Comment:
A second con: if there is a change in the cloud that hasn't yet been synced, the full apply will blow it away and then the follow-up sync step will do nothing (in fact, there is no point in doing the sync at all, anymore, because the DB becomes the sole source-of-truth and we only sync during install).